lib/backends/webgl/inference-handler.ts

```diff
@@ -5,105 +5,151 @@ import {InferenceHandler} from '../../backend';
+    const texture = this.textureHelper.createTextureFromLayout(dataType, layout, data, usage);
+    return this.createTextureDataFromTexture(layout, dataType, texture, tensor);
+  }
+
+  /**
+   * Create a TextureData object, using the given texture.
+   * This function does not create new texture. Usually used in scenarios using texture sharing. (eg. Reshape)
+   * @param dataType the tensor data type
+   * @param texture the WebGLTexture object to share
+   * @param tensorId the tensor ID of the shared tensor data
+   */
+  createSharedTextureData(layout: TextureLayout, dataType: Tensor.DataType, texture: WebGLTexture, tensorId: Tensor.Id):
+      TextureData {
+    return this.createTextureDataFromTexture(layout, dataType, texture, undefined, tensorId);
+  }
+
+  private createTextureDataFromTexture(
+      layout: TextureLayout, dataType: Tensor.DataType, texture: WebGLTexture, tensor?: Tensor, tensorId?: Tensor.Id) {
+    const textureData: TextureData = {
+      ...layout,
+      tensor: tensor ||
+          new Tensor(
+                  layout.unpackedShape, dataType,
+                  (id: Tensor.Id) => {
+                    return this.readTexture(textureData);
+                  },
+                  undefined, undefined, tensorId),
+      texture
+    };
+    this.setTextureData(textureData.tensor.dataId, textureData);
+    return textureData;
   }
-  getTensor(td: TextureData): Tensor {
-    let tensor: Tensor|undefined;
-    tensor = this.textureToTensor.get(td);
-    if (!tensor) {
-      Logger.verbose('InferenceHandler', `Creating new Tensor from texture data: [${td.unpackedShape}]`);
-      /**
-       * We're creating a Tensor without converting data from Texture onto CPU
-       * Instead we're passing a closure which is only executed if Tesor.data is accessed
-       * This allows for the execution of the graph without paying the penalty of
-       * data movement from GPU to CPU
-       */
-      tensor = new Tensor(td.unpackedShape, td.dataType, (id: Tensor.Id) => {
-        return this.readTexture(td);
-      });
-      this.setTextureData(tensor, td);
+
+  getTextureData(tensorId: Tensor.Id): TextureData|undefined {
+    return this.session.isInitializer(tensorId) ? this.session.getTextureData(tensorId) :
+                                                  this.textureDataCache.get(tensorId);
+  }
+  setTextureData(tensorId: Tensor.Id, td: TextureData): void {
+    if (this.session.isInitializer(tensorId)) {
+      this.session.setTextureData(tensorId, td);
     } else {
-      Logger.verbose('InferenceHandler', `Retrieving Tensor from cache for:[${td.unpackedShape}]`);
+      this.textureDataCache.set(tensorId, td);
     }
-    return tensor;
   }
+
+  /**
+   * Create a TextureLayout object from a tensor. If a related texture data is found, returns the cached texture layout.
+   */
   getOrCreateTextureLayout(tensor: Tensor, channels = 1, unpackedShape?: ReadonlyArray<number>): TextureLayout {
```

Comment:
Nit: This name is slightly misleading. I would expect that `getOrCreateTexturelayout()` to only focus on Layout  creation. There is a step that looks into the cache preceeding it. Any possible better names for this function ?

Reply:
This is what the function is doing: try to get (find from cache) the texture layout of the given tensor; if failed to find then create a new one.. so from the naming perspective I cannot find a better name for it. This is the same for `getOrCreateTextureData`.